test edge commit image

yih-redhat opened the pull request:
This PR contains:

1. Disabled subscription for rhel8.4 runners
2. Added test script to test edge commit image

@atodorov Can you please help to review this one? Appreciate if you can help to start to run this PR, it passed locally but it's the first time to run in image-builder CI, I need to run it early and fix the environment problems.

atodorov commented on .gitlab-ci.yml:
```diff
@@ -42,6 +42,21 @@ Test EL8:
+  stage: test
+  extends: .terraform
+  variables:
+    EXTRA_REPO_PATH_SEGMENT: "gitlab/"
```
Unrelated but EXTRA_REPO_PATH_SEGMENT was removed in osbuild-composer so I suggest removing it here as well.

yih-redhat replied:
Thanks, removed.

atodorov commented on test/cases/edge.sh:
```diff
@@ -0,0 +1,400 @@
+            OSTREE_REF="rhel/8/${ARCH}/edge"
+            OS_VARIANT="rhel8-unknown"
+            DISTRO="rhel-84"
+            BOOT_LOCATION="http://download.devel.redhat.com/rel-eng/rhel-8/RHEL-8/latest-RHEL-8.4.0/compose/BaseOS/x86_64/os/";;
```
Use ${ARCH} in the URL too b/c it will fail on non x86_64 targets.

atodorov replied:
^^^ still unresolved.

yih-redhat replied:
thanks, resolved.

atodorov reviewed (changes requested):
Minor changes requested, need to see what the results from GitLab CI are.

achilleas-k commented:
Looks good but, as discussed in meeting just now, we decided to put the kickstart file on the ISO using `mkksiso` instead of injecting it during `virt-install`.

yih-redhat commented:
> Looks good but, as discussed in meeting just now, we decided to put the kickstart file on the ISO using `mkksiso` instead of injecting it during `virt-install`.

Yes.
This PR only includes the test case for commit image type. And the test case for installer ISO image type is ready, but will be submitted in next PR. (of course I need to change to use mkksiso now)
Because this is the first time I run test in image-builder CI, I prefer to submit a small PR first, and if everything is fine, more test cases will come in.

atodorov reviewed (changes requested):
DEBUG:     Installed (but unpackaged) file(s) found:
DEBUG:    /usr/share/tests/image-builder/edge/check_ostree.yaml
DEBUG:    /usr/share/tests/image-builder/keyring/id_rsa
DEBUG:    /usr/share/tests/image-builder/keyring/id_rsa.pub

yih-redhat commented:
Hi @atodorov ,

Gitlab job failed at schutzbot/build.sh, looks like the runner machine is registered already and trying to register again.

$ schutzbot/build.sh
package epel-release is not installed
📦 Setting up EPEL repository
Verifying...                          ########################################
warning: /tmp/epel.rpm: Header V4 RSA/SHA256 Signature, key ID 2f86d6a1: NOKEY
Preparing...                          ########################################
Updating / installing...
epel-release-8-11.el8                 ########################################
+-------------------------------------------+
   System Status Details
+-------------------------------------------+
Overall Status: Unknown
System Purpose Status: Unknown
WARNING
The yum/dnf plugins: /etc/dnf/plugins/subscription-manager.conf, /etc/dnf/plugins/product-id.conf were automatically enabled for the benefit of Red Hat Subscription Management. If not desired, use "subscription-manager config --rhsm.auto_enable_yum_plugins=0" to block this behavior.
🪙 Registering RHEL instance
Registering to: subscription.rhsm.redhat.com:443/subscription
The system has been registered with ID: 9fc6e7ce-f4a0-42ce-a085-2edf39c46b1a
The registered system name is: rhel-8-4
Error: 'codeready-builder-for-rhel-8-x86_64-rpms' does not match a valid repository ID. Use "subscription-manager repos --list" to see valid repositories.

yih-redhat commented:
> DEBUG: Installed (but unpackaged) file(s) found:
> DEBUG: /usr/share/tests/image-builder/edge/check_ostree.yaml
> DEBUG: /usr/share/tests/image-builder/keyring/id_rsa
> DEBUG: /usr/share/tests/image-builder/keyring/id_rsa.pub

Fixed, thanks.

atodorov commented on test/cases/edge.sh:
```diff
@@ -0,0 +1,400 @@
+    },
+    "image_requests": [
+      {
+        "architecture": "x86_64",
```
Use $(ARCH) here too.

yih-redhat replied:
fixed, thanks.

yih-redhat commented on test/cases/edge.sh:
```diff
@@ -0,0 +1,400 @@
+            OSTREE_REF="rhel/8/${ARCH}/edge"
+            OS_VARIANT="rhel8-unknown"
+            DISTRO="rhel-84"
+            BOOT_LOCATION="http://download.devel.redhat.com/rel-eng/rhel-8/RHEL-8/latest-RHEL-8.4.0/compose/BaseOS/${ARCH}/os/";;
```
ok, changed to 8.5

atodorov replied:
Any reason to revert to 8.4 after all the previous work and changes to use 8.5 ?

yih-redhat replied:
Achilleas confirmed image-builder service can only run on rhel8.4 for edge Rest API.

yih-redhat commented on .gitlab-ci.yml:
```diff
@@ -42,6 +42,19 @@ Test EL8:
+  parallel:
+    matrix:
+      - RUNNER:
+          - openstack/rhel-8.4-x86_64
```
Hmmm, I changed to 8.5, and got this error.

Turn off this advice by setting config variable advice.detachedHead to false
HEAD is now at aea553f add centos stream 8 image in openstack
cat: /tmp/custom-executor400308399/tmp.zsdZqRoGvq/openstack/rhel-8.5-x86_64/config.json: No such file or directory
Error handling -chdir option: chdir /home/centos/jobs/openstack/rhel-8.5-x86_64/1412050914/openstack/rhel-8.5-x86_64: no such file or directory
cat: /home/centos/jobs/openstack/rhel-8.5-x86_64/1412050914/ip: No such file or directory
Host  not found in /home/centos/.ssh/known_hosts

atodorov replied:
Is there a reason to revert this back to 8.4 after previously switching to 8.5 ?

yih-redhat replied:
Achilleas confirmed image-builder cannot run on rhel 8.5, so I changed it back.

atodorov reviewed (changes requested):
Minor changes requested. However the status box under this PR says;
`This branch is out-of-date with the base branch` which probably means you have to rebase on top of the latest `main` branch.

yih-redhat commented on .gitlab-ci.yml:
```diff
@@ -42,6 +42,19 @@ Test EL8:
         RUNNER:
           - aws/rhel-8-x86_64
 
+Test Edge 8.4:
```
fixed, thanks.

yih-redhat commented on test/cases/edge.sh:
```diff
@@ -0,0 +1,400 @@
+            IMAGE_TYPE="rhel-edge-commit"
+            OSTREE_REF="rhel/8/${ARCH}/edge"
+            OS_VARIANT="rhel8-unknown"
+            DISTRO="rhel-84"
```
updated to 85, thanks.

yih-redhat commented on test/cases/edge.sh:
```diff
@@ -0,0 +1,400 @@
+
+    # Set os-variant and boot location used by virt-install.
+    case "${ID}-${VERSION_ID}" in
+        "rhel-8.4")
```
yes, updated

yih-redhat commented:
Hi @atodorov ,

I rebased with latest main branch, and updated test cases to 8.5, but still see this error:

Turn off this advice by setting config variable advice.detachedHead to false
HEAD is now at aea553f add centos stream 8 image in openstack
cat: /tmp/custom-executor910393740/tmp.kAl3xBd4UN/openstack/rhel-8.5-x86_64/config.json: No such file or directory
Error handling -chdir option: chdir /home/centos/jobs/openstack/rhel-8.5-x86_64/1412135674/openstack/rhel-8.5-x86_64: no such file or directory
cat: /home/centos/jobs/openstack/rhel-8.5-x86_64/1412135674/ip: No such file or directory
Host  not found in /home/centos/.ssh/known_hosts
ERROR: Job failed: exit status 1

yih-redhat commented:
As openstack/rhel-8.4-x86_64 can not subscribe, I disabled subscription for it, and fixed mock build failures because of not subscribed. For image-builder service, I also replaced the repo link from CDN to internal repo link. Now all checks passed and this PR is green.

@atodorov can you please close the 'request change' that blocks the merge?

atodorov commented on schutzbot/build.sh:
```diff
@@ -19,8 +19,17 @@ if [[ $ID == rhel ]] && ! rpm -q epel-release; then
+#     greenprint "🪙 Registering RHEL instance"
+#     sudo chmod +x "$RHN_REGISTRATION_SCRIPT"
+#     sudo "$RHN_REGISTRATION_SCRIPT"
+# fi
```
This looks like a left over given the changes a few lines below.

yih-redhat replied:
Hmm... don't know why it looks like this. Will try to fix it.

yih-redhat replied:
Fixed, thanks.

atodorov commented on schutzbot/build.sh:
```diff
@@ -19,8 +19,17 @@ if [[ $ID == rhel ]] && ! rpm -q epel-release; then
+# Currently openstack/rhel-8.4-x86_64 cannot subcribe, disable subscription.
+# TODO: remove this after openstack/rhel-8.4-x86_64 can subscribe
+echo "Do not subscribe for rhel84."
+if [[ $ID == "rhel" && $VERSION_ID == "8.3" && -n "${RHN_REGISTRATION_SCRIPT:-}" ]] && ! sudo subscription-manager status; then
```
I am wondering if this condition can be based not on the specific OS version but the fact that it is a BETA release ?

yih-redhat replied:
Agree, it should be based on BETA release.
I have no idea how to check if RHEL is BETA, I have created a jira task for it.
https://issues.redhat.com/browse/COMPOSER-1034

atodorov replied:
The information is already contained inside the /etc/os-release file. We don't need a separate ticket for that.

yih-redhat replied:
Fixed, thanks.

atodorov commented on schutzbot/deploy.sh:
```diff
@@ -123,6 +123,12 @@ for SOURCE in $(sudo composer-cli sources list); do
+# Can not pull from remote repo, use my own user/pass here for test.
+# Should not expose the account here, will replace them with test team account
+# Remove this after openstack/rhel-8.4-x86_64 can subscribe
+sudo podman login docker.io -u yihredhat -p p0o9i8u7Ginker1!
```
Login/Password exposed in plain text. Change those immediately!

yih-redhat replied:
Because subscription is disabled, when start Postgres container, it failed to connect to docker.io, that's why I asked is there a test account that I can use. If not, I need your help to set login/password as project env variables that I can use them.

yih-redhat replied:
Fixed, thanks.

atodorov reviewed (dismissed):
- Posted some questions and code review.
- This PR contains exposed credentials so change those immediately and don't expose them in source code

What's the reason for the `podman login` command?

ondrejbudai commented:
Hmm, I've just realized if this is the right way. Image builder is not meant to be run on pre-release RHEL (`openstack/rhel-8.4-x86_64` image is a pre-release one). Shouldn't we run this test on a released version instead?

henrywang commented on test/data/repositories/rhel-84.json:
```diff
@@ -0,0 +1,15 @@
+    "x86_64": {
+      "image_types": [ "ami", "vhd", "rhel-edge-commit", "rhel-edge-installer" ],
+      "repositories": [{
+        "baseurl": "http://download.devel.redhat.com/rhel-8/nightly/RHEL-8/latest-RHEL-8.5/compose/BaseOS/x86_64/os/"
```
The repo does not pin to rhel 8.4 internal repo, but rhel 8.5. Your image is not RHEL 8.4, but RHEL 8.5.
@yih-redhat You have to re-run the test after you pin image-builder repo to rhel 8.4.

yih-redhat replied:
Hmm, this is the repo I got from Achilleas...
I will change to 8.4, and see if it works.

yih-redhat replied:
changed to 8.4, and passed. 
but another test job failed, which has nothing to do with my change...
will try to rerun.

yih-redhat replied:
rerun this pr, passed.

henrywang reviewed (dismissed):
LGTM. Thanks @yih-redhat

ondrejbudai commented:
I wonder if we can change the runner to `openstack/rhel-8-x86_64` instead of `openstack/rhel-8.4-x86_64`.

There's a quite big difference between these two:
- `openstack/rhel-8-x86_64` is a RHEL 8.3 GA image that's pre-subscribed.
- `openstack/rhel-8.4-x84_86` is a random RHEL 8.4 nightly. Due to its beta nature, it's basically impossible to subscribe properly.

I've started a pipeline with the GA image here: https://gitlab.com/osbuild/ci/image-builder/-/pipelines/335627610
If this passes, we might want to switch to this image and remove all the workarounds.

yih-redhat commented:
> I wonder if we can change the runner to `openstack/rhel-8-x86_64` instead of `openstack/rhel-8.4-x86_64`.
> 
> There's a quite big difference between these two:
> 
> * `openstack/rhel-8-x86_64` is a RHEL 8.3 GA image that's pre-subscribed.
> * `openstack/rhel-8.4-x84_86` is a random RHEL 8.4 nightly. Due to its beta nature, it's basically impossible to subscribe properly.
> 
> I've started a pipeline with the GA image here: https://gitlab.com/osbuild/ci/image-builder/-/pipelines/335627610
> If this passes, we might want to switch to this image and remove all the workarounds.

Hi Ondrej,

Thanks for your update.

My test scripts can only run on rhel 8.4, because the  edge image type - installer ISO, is not supported on rhel 8.3.

Thanks,
Yi

yih-redhat commented:
@atodorov Hi, can you please approve the change request and merge this pr?

ondrejbudai commented:
@yih-redhat It will be fine, you can still specify `DISTRO=rhel-84` even when not running the test on RHEL 8.4. Both `image-builder` and `osbuild-composer` can build images for different distributions than that they are running on.

ondrejbudai commented:
See my commit: https://github.com/osbuild/image-builder/commit/cb10a794e2770e8f05f274da8e457805c80f6afa

and its result: https://gitlab.com/osbuild/ci/image-builder/-/jobs/1417915941

btw I don't think we need to differentiate between RHEL versions in `image-builder`, so I would just remove the whole `case` statement from the script.

yih-redhat commented:
> See my commit: [cb10a79](https://github.com/osbuild/image-builder/commit/cb10a794e2770e8f05f274da8e457805c80f6afa)
> 
> and its result: https://gitlab.com/osbuild/ci/image-builder/-/jobs/1417915941
> 
> btw I don't think we need to differentiate between RHEL versions in `image-builder`, so I would just remove the whole `case` statement from the script.

Thanks Ondrej, I will update this PR as you did in your commit, and I will remove the 'case' statement.

yih-redhat commented:
the GCP test cases in api.sh failed randomly, which made gitlab CI failed. @atodorov Do you know who is the owner of this case?

henrywang reviewed (changes requested):
@yih-redhat From the https://gitlab.com/osbuild/ci/image-builder/-/jobs/1419639909#L4797, you're running test on RHEL 8.3 VM. `rhel-edge-commit` and `rhel-edge-installer` are for RHEL 8.4, that means osbuild* have to be installed into RHEL 8.4 VM, not 8.3. Could you please check your running environment?

Gundersanne commented:
> @yih-redhat From the https://gitlab.com/osbuild/ci/image-builder/-/jobs/1419639909#L4797, you're running test on RHEL 8.3 VM. `rhel-edge-commit` and `rhel-edge-installer` are for RHEL 8.4, that means osbuild* have to be installed into RHEL 8.4 VM, not 8.3. Could you please check your running environment?

I think running the test on rhel8.3 is fine. Image Builder (the service), runs inside of a ubi container (also during tests), and I'm not even sure what the host machines run in CRC. Which is why for all the tests here we just picked GA.

The only way this will break is if the chosen osbuild/composer commits won't install anymore on the machine, at which point we'd switch over all tests to a new host. I get that this is not the ideal, but as soon as we have rhel 8.4 GA machines which subscribe easily, we can run them on that.

henrywang commented:
Thanks @Gundersanne! I found the osbuild and osbuild-composer are installed in host from `deploy.sh`https://github.com/osbuild/image-builder/blob/3b7a584a2a48a648e7e2524fe3bc3257a7ebe303/schutzbot/deploy.sh#L110. So I think we should use rhel 8.4 VM.

henrywang reviewed (approved):
LGTM.
